### PR TITLE
Dev/alex/auth service 2.0 : "auth/roles/list/:role"

### LINKED ISSUE #12: Auth Service 2.0
What's optimal?

### PR DESCRIPTION
**Pull request by aletya**
In /src/services/auth/auth-lib.ts, added a function getUsersWithRole which takes in the role to filter for and returns a string[] of ids.
In src/services/auth/auth-router.ts, added route "auth/roles/list/:role", which get all users that have a certain role, returning a string[] of ids.

### CODE REVIEW

File: src/services/auth/auth-lib.ts [hunk 1]
```diff
@@ -316,3 +316,27 @@ export function getDevice(kv?: string): string {
...
+
+
+/**
+ * Get all id of users that have a particular role within the database.
+ * @param role role that we want to filter for
+ * @returns Promise<string[]> - if valid, then contains array of user w/ role. If invalid, then contains "Unknown Error".
+ */
+export async function getUsersWithRole(role : string): Promise<string[]> {
+
+	// Makse a reference to the roles collection
```
Comment (AydanPirani) [inline src/services/auth/auth-lib.ts]: Typo

File: src/services/auth/auth-lib.ts [hunk 2]
```diff
@@ -316,3 +316,27 @@ export function getDevice(kv?: string): string {
...
+	const collection: Collection = databaseClient.db(Constants.AUTH_DB).collection(AuthDB.ROLES);
+
+	//now make a mongodb query that iterates thru and check which ones have the role
```
Comment (AydanPirani) [inline src/services/auth/auth-lib.ts]: Capitalize comments please, and comments should be a bit more formal... i.e "thru" to "through"

File: src/services/auth/auth-lib.ts [hunk 3]
```diff
@@ -316,3 +316,27 @@ export function getDevice(kv?: string): string {
...
+	const queryCriteria : { roles: { $in: string[] } } = { roles: { $in: [role] } };
```
Comment (AydanPirani) [inline src/services/auth/auth-lib.ts]: `{ roles: { $in: string[] } }` --> Create/use a pre-defined type instead. I think you can use a `Filter<Document` instead, which makes this cleaner

File: src/services/auth/auth-lib.ts [hunk 4]
```diff
@@ -316,3 +316,27 @@ export function getDevice(kv?: string): string {
...
+
+	//array of users as MongoDb schema that have role as one of its roles
+	const result : RolesSchema[] = await collection.find(queryCriteria).toArray() as RolesSchema[];
+	//array of strings for id, will be the return value of this funciton
+	const idArray : string[] = (result).map((user : RolesSchema) => {
```
Comment (AydanPirani) [inline src/services/auth/auth-lib.ts]: Why parenthesis around result?

File: src/services/auth/auth-router.ts [hunk 5]
```diff
@@ -312,6 +312,37 @@ authRouter.get("/roles/", strongJwtVerification, async (_: Request, res: Respons
...
+ * @apiSuccess (200: Success) {String[]} Array of ids of users w/ the specified role.
+ * @apiSuccessExample Example Success Response:
+ * 	HTTP/1.1 200 OK
+ *	{
+ *		"data" : ["github44122133", "github22779056", "github5997469", "github98075854"]
+ * 	}
+ *
+ * @apiUse strongVerifyErrors
+ */
+authRouter.get("/roles/list/:role", async (req: Request, res: Response) => {
```
Comment (AydanPirani) [inline src/services/auth/auth-router.ts]: `role` should be capitalized in the query params, so :role -> :ROLE

File: src/services/auth/auth-router.ts [hunk 6]
```diff
@@ -312,6 +312,37 @@ authRouter.get("/roles/", strongJwtVerification, async (_: Request, res: Respons
...
+	const role: string | undefined = req.params.role?.toString();
+
+	//returns error if role parameter is empty
+	if (!role) {
+		return res.status(Constants.BAD_REQUEST).send({ error: "Role parameter is missing" });
```
Comment (AydanPirani) [inline src/services/auth/auth-router.ts]: `error: InvalidParams`

**Review by AydanPirani — CHANGES_REQUESTED** [review]
Left a few small nits, should be good to go afterwards!

**Comment by aletya** [comment]
Fixed pr comments

File: src/services/auth/auth-router.ts [hunk 7]
```diff
@@ -312,6 +312,37 @@ authRouter.get("/roles/", strongJwtVerification, async (_: Request, res: Respons
...
+ * @apiSuccessExample Example Success Response:
+ * 	HTTP/1.1 200 OK
+ *	{
+ *		"data" : ["github44122133", "github22779056", "github5997469", "github98075854"]
+ * 	}
+ *
+ * @apiUse strongVerifyErrors
+ */
+authRouter.get("/roles/list/:ROLE", async (req: Request, res: Response) => {
+	const ROLE: string | undefined = req.params.ROLE?.toString();
```
Comment (AydanPirani) [inline src/services/auth/auth-router.ts:331]: sorry for the confusion! role in your code should be lowercase, but :ROLE should be uppercase.

So something like `const role: string | undefined = req.params.ROLE;

Comment (AydanPirani) [inline src/services/auth/auth-router.ts]: Sorry for the confusion! 

role in code should be lowercase, but caps in the route.

For example: 

`const role: string | undefined = req.params.ROLE;` (also dw abt the toString function I think)

File: src/services/auth/auth-lib.ts [hunk 8]
```diff
@@ -316,3 +316,31 @@ export function getDevice(kv?: string): string {
...
+	/*type Filter = {
+		roles: { $in: string[] };
+	};*/
+
+	//Makes a mongodb query that iterates through the roles array for each user and selects ones that have wanted role
+	const queryCriteria : Filter<Document> = { roles: { $in: [role] } };
+
+	//array of users as MongoDb schema that have role as one of its roles
+	const result : RolesSchema[] = await collection.find(queryCriteria).toArray() as RolesSchema[];
+	//array of strings for id, will be the return value of this funciton
```
Comment (AydanPirani) [inline src/services/auth/auth-lib.ts]: Also capitalize all these comments

File: src/services/auth/auth-lib.ts [hunk 9]
```diff
@@ -316,3 +316,31 @@ export function getDevice(kv?: string): string {
...
+ * @returns Promise<string[]> - if valid, then contains array of user w/ role. If invalid, then contains "Unknown Error".
+ */
+export async function getUsersWithRole(role : string): Promise<string[]> {
+
+	// Makes a reference to the roles collection
+	const collection: Collection = databaseClient.db(Constants.AUTH_DB).collection(AuthDB.ROLES);
+	
+	/*type Filter = {
+		roles: { $in: string[] };
+	};*/
```
Comment (AydanPirani) [inline src/services/auth/auth-lib.ts]: Remove commented-out code

File: src/services/auth/auth-router.ts [hunk 10]
```diff
@@ -312,6 +312,37 @@ authRouter.get("/roles/", strongJwtVerification, async (_: Request, res: Respons
...
+
+	//returns error if role parameter is empty
+	if (!ROLE) {
+		return res.status(Constants.BAD_REQUEST).send({ error: "InvalidParams" });
+	}
+
+	return await getUsersWithRole(ROLE).then((users: string[]) => {
+		return res.status(Constants.SUCCESS).send({ data: users });
```
Comment (AydanPirani) [inline src/services/auth/auth-router.ts]: Super small nit, but change `data: users` to `userIds: users`

**Review by AydanPirani — CHANGES_REQUESTED** [review]
Few more changes, then we're good to go!

**Comment by AydanPirani** [comment]
Looks good to me! Thanks for working on this :)

**Comment by aletya** [comment]
Yay first pr/push! Thanks for the comments

**Review by AydanPirani — APPROVED** [review]
Feel free to merge whenever :)

**Pull request merged**